ITE/10/README.adoc

```diff
@@ -0,0 +1,278 @@
+  indicate the products of the step. Eg. in-toto links, SLSA Provenance.
+* informational predicates: these predicates record contextual attributes about
+  one or more artifacts. The operations they represent consume artifacts as
+  materials (recorded as the subject of such attestations) and produce
+  attributes about them as the outcome. Eg. test results, reviews.
+
+Note that transformational predicates may also record certain information about
+certain attributes. For example, the environment field in in-toto links can
+include such information.
+
+With this representation of different predicate types, step declarations in
+layouts can _chain_ different operations represented by distinct predicate types
+together, a fundamental feature of in-toto. For example, a code review step
+represented by a review predicate type that follows a checkout step that fetches
+source code represented by an in-toto link can use in-toto's artifact rules to
+ensure the review was performed against the sources fetched by the checkout
+step.
+
+=== Setting Predicate Type Expectations for Steps and Inspections
+
+Currently, step declarations have the following schema.
+
+```yaml
+name: string
+threshold: int
+expectedMaterials: list of artifact rules
+expectedProducts: list of artifact rules
+pubkeys: list of authorized key IDs
+expectedCommand: list of strings
+```
+
+This ITE proposes updating this schema to incorporate predicate types.
+
+```yaml
+name: string
+command: list of strings
+expectedMaterials: list of artifact rules
+expectedPredicates: list of expectedStepPredicates
+```
+
+Each `expectedStepPredicate` object has the following schema.
+
+```yaml
+predicateType: string
+expectedProducts: list of artifact rules
+functionaries: list of authorized functionary key IDs
+threshold: int
+```
+
+The `predicateType` field identifies the type of predicate expected for the
+step. `expectedProducts` is moved to being predicate specific as informational
+predicates may be generated alongside transformational predicates for the same
+step. This means that product rules do not generalize across all of the step's
+predicates. However, `expectedMaterials` is retained at the step level as all
+predicates generated for a step are expected to be for the same set of
+artifacts. `pubkeys` is updated to `functionaries` to generalize signing
+semantics. Finally, `threshold` applies per predicate type rather than the step
+as a whole.
+
+Similarly, inspections are also updated to support generating different
+predicate types.
+
+```yaml
+name: string
+command: list of strings
+expectedMaterials: list of artifact rules
+expectedPredicates: list of expectedInspectionPredicates
+```
+
+Each `expectedInspectionPredicate` object has the following schema.
+
+```yaml
+predicateType: string
+expectedProducts: list of artifact rules
+```
+
+As the attestations generated during inspection executions are not signed,
+`functionaries` and `threshold` are omitted.
```

Comment:
Who consumes inspection attestations? I'd imagine some downstream/third-party verifier or auditor might deem the output of an inspection to be more trustworthy if it's signed.

Reply:
in-toto's v1 verification workflow uses inspection links. I think this was basically so that artifact rules could be used to connect inspection results to other steps. I think the classic example is an unzip inspection which has a MATCH rule to ensure the contents of the archive match the products of the build step and so on. Nothing really stops an inspection link / attestation from being signed and distributed IMO, but we shouldn't roll that into the verification workflow because we don't expect the verifier to be capable of any signing in the first place (and indeed the verifier isn't really associated with an identity either).